metadata-ingestion/src/datahub/ingestion/source/hex/api.py

```diff
@@ -0,0 +1,394 @@
+    """https://learn.hex.tech/docs/api/api-reference"""
+
+    def __init__(
+        self,
+        token: str,
+        report: HexApiReport,
+        base_url: str = HEX_API_BASE_URL_DEFAULT,
+        page_size: int = HEX_API_PAGE_SIZE_DEFAULT,
+    ):
+        self.token = token
+        self.base_url = base_url
+        self.report = report
+        self.page_size = page_size
+
+    def _list_projects_url(self):
+        return f"{self.base_url}/projects"
+
+    def _auth_header(self):
+        return {"Authorization": f"Bearer {self.token}"}
+
+    def fetch_projects(
+        self,
+        include_components: bool = True,
+        include_archived: bool = False,
+        include_trashed: bool = False,
+    ) -> Generator[Union[Project, Component], None, None]:
+        """Fetch all projects and components
+
+        https://learn.hex.tech/docs/api/api-reference#operation/ListProjects
+        """
+        params = {
+            "includeComponents": include_components,
+            "includeArchived": include_archived,
+            "includeTrashed": include_trashed,
+            "includeSharing": True,
+            "limit": self.page_size,
+            "after": None,
+            "before": None,
+            "sortBy": "CREATED_AT",
+            "sortDirection": "ASC",
+        }
+        yield from self._fetch_projects_page(params)
+
+        while params["after"]:
+            yield from self._fetch_projects_page(params)
+
+    def _fetch_projects_page(
+        self, params: Dict[str, Any]
+    ) -> Generator[Union[Project, Component], None, None]:
+        logger.debug(f"Fetching projects page with params: {params}")
+        self.report.fetch_projects_page_calls += 1
+        try:
+            response = requests.get(
+                url=self._list_projects_url(),
+                headers=self._auth_header(),
+                params=params,
+                timeout=30,
+            )
+            response.raise_for_status()
+
+            api_response = HexApiProjectsListResponse.parse_obj(response.json())
+            logger.info(f"Fetched {len(api_response.values)} items")
+            params["after"] = (
+                api_response.pagination.after if api_response.pagination else None
+            )
+
+            self.report.fetch_projects_page_items += len(api_response.values)
+
+            for item in api_response.values:
+                try:
+                    ret = self._map_data_from_model(item)
+                    yield ret
+                except Exception as e:
+                    self.report.warning(
+                        title="Incomplete metadata",
+                        message="Incomplete metadata because of error mapping item",
+                        context=str(item),
+                        exc=e,
+                    )
+        except ValidationError as e:
+            self.report.failure(
+                title="API response parsing error",
+                message="Error parsing API response",
```

Comment:
can we make these errors more specific - e.g. which api response failed to parse, and what will that actually impact

As an end user, if I saw "API response parsing error" I wouldn't really know what to do about it / how much I should care

Reply:
addressed in commit 2a2d590faadd49f9c505e08c7542e268e58cf818